File: ios/ExpoEzvizModule.swift
import ExpoModulesCore
import EZOpenSDKFramework

public class ExpoEzvizModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    Name("ExpoEzviz")


    // Defines event names that the module can send to JavaScript.
    Events("onLoad")

    AsyncFunction("initSDK") { (appKey: String, apiUrl: String?) in
      // Disable P2P before initializing the SDK
      EZOpenSDK.enableP2P(false)
      if let url = apiUrl, !url.isEmpty {
        // set debug mode first
        EZOpenSDK.setDebugLogEnable(true)
        EZOpenSDK.initLib(withAppKey: appKey, url: url, authUrl: "")
      } else {
        EZOpenSDK.initLib(withAppKey: appKey)
      }
      print("EZOpenSDK version: \(EZOpenSDK.getVersion())")
    }

    AsyncFunction("createPlayer") { (view: ExpoEzvizView) in
        view.createPlayer()
    }

    AsyncFunction("setAccessToken") { (accessToken: String) in
      print("EzvizModule: Setting Access Token")
      EZOpenSDK.setAccessToken(accessToken)
    }

    AsyncFunction("searchRecordFileFromDevice") { (deviceSerial: String, cameraNo: Int, beginTimeString: String, endTimeString: String, promise: Promise) in
      let dateFormatter = DateFormatter()
      dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
      dateFormatter.timeZone = TimeZone(abbreviation: "UTC")
      guard let beginTime = dateFormatter.date(from: beginTimeString),
            let endTime = dateFormatter.date(from: endTimeString) else {
          promise.reject("date-format-error", "Could not parse date strings. Expected format: yyyy-MM-dd HH:mm:ss")
          return
      }

      let timeInterval = endTime.timeIntervalSince(beginTime)
      if timeInterval > 24 * 60 * 60 {
          promise.reject("query-window-error", "The time window for querying records cannot exceed 24 hours.")
          return
      }

      EZOpenSDK.searchRecordFile(fromDevice: deviceSerial, cameraNo: cameraNo, beginTime: beginTime, endTime: endTime) { deviceRecords, error in
        if let error = error {
          promise.reject("search-error", error.localizedDescription)
          return
        }
 
        let serializableRecords = deviceRecords.map { record -> [String: Any?] in
            guard let record = record as? EZDeviceRecordFile else { return [:] }
            return [
              "startTime": record.startTime.timeIntervalSince1970 * 1000, // JS uses milliseconds
              "stopTime": record.stopTime.timeIntervalSince1970 * 1000
            ]
        }
        promise.resolve(serializableRecords)
      }
    }

    AsyncFunction("addDevice") { (deviceSerial: String, verifyCode: String, promise: Promise) in
      EZOpenSDK.addDevice(deviceSerial, verifyCode: verifyCode) { error in
        if let error = error {
          promise.reject("add-device-error", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      }
    }

    AsyncFunction("deleteDevice") { (deviceSerial: String, promise: Promise) in
      EZOpenSDK.deleteDevice(deviceSerial) { error in
        if let error = error {
          promise.reject("delete-device-error", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      }
    }

    AsyncFunction("setDeviceName") { (deviceName: String, deviceSerial: String, promise: Promise) in
      EZOpenSDK.setDeviceName(deviceName, deviceSerial: deviceSerial) { error in
        if let error = error {
          promise.reject("set-device-name-error", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      }
    }

    AsyncFunction("setDeviceEncryptStatus") { (deviceSerial: String, verifyCode: String, encryptStatus: Bool, promise: Promise) in
      EZOpenSDK.setDeviceEncryptStatus(deviceSerial, verifyCode: verifyCode, encrypt: encryptStatus) { error in
        if let error = error {
          promise.reject("set-encrypt-status-error", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      }
    }

    AsyncFunction("setDefence") { (deviceSerial: String, defenceType: Int, promise: Promise) in
      guard let type = EZDefenceStatus(rawValue: defenceType) else {
        promise.reject("invalid-defence-type", "Invalid defence type. Use 0 (Off/Sleep), 1 (On), 8 (At Home), or 16 (Outer).")
        return
      }

      EZOpenSDK.setDefence(type, deviceSerial: deviceSerial) { error in
        if let error = error {
          promise.reject("set-defence-error", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      }
    }

    AsyncFunction("getDeviceInfo") { (deviceSerial: String, promise: Promise) in
      EZOpenSDK.getDeviceInfo(deviceSerial) { deviceInfo, error in
        if let error = error {
          promise.reject("get-device-info-error", error.localizedDescription)
          return
        }

        // Manually create a dictionary with the desired properties to avoid KVC crashes.
        // Some of the elements are swallowed by the EZViz SDK.
        let deviceInfoDict: [String: Any?] = [
            // "addTime": deviceInfo.addTime,
            // "alarmSoundMode": deviceInfo.alarmSoundMode,
            "cameraNum": deviceInfo.cameraNum,
            "category": deviceInfo.category,
            "defence": deviceInfo.defence,
            "detectorNum": deviceInfo.detectorNum,
            "deviceCover": deviceInfo.deviceCover,
            "deviceName": deviceInfo.deviceName,
            "deviceSerial": deviceInfo.deviceSerial,
            "deviceType": deviceInfo.deviceType,
            "deviceVersion": deviceInfo.deviceVersion,
            "isEncrypt": deviceInfo.isEncrypt,
            // "offlineNotify": deviceInfo.offlineNotify,
            "status": deviceInfo.status,
            // "supportExtShort": deviceInfo.supportExtShort
        ]
        promise.resolve(deviceInfoDict)
      }
    }

    AsyncFunction("destoryPlayer") { (view: ExpoEzvizView) in
        // note typo from EZOpenSDK
        view.destoryPlayer()
    }


    // Enables the module to be used as a native view. Definition components that are accepted as part of the
    // view definition: Prop, Events.
    View(ExpoEzvizView.self) {
      // Defines a setter for the `url` prop.
      Prop("deviceSerial") { (view: ExpoEzvizView, prop: String?) in view.deviceSerial = prop }
      Prop("cameraNo") { (view: ExpoEzvizView, prop: Int) in view.cameraNo = prop }
      Prop("accessToken") { (view: ExpoEzvizView, prop: String?) in view.accessToken = prop }
      Prop("verifyCode") { (view: ExpoEzvizView, prop: String?) in view.verifyCode = prop }
      Prop("autoplay") { (view: ExpoEzvizView, prop: Bool?) in view.autoplay = prop ?? false }
      Prop("defaultSoundOn") { (view: ExpoEzvizView, prop: Bool?) in view.setDefaultSoundOn(prop) }

      Events("onLoad", "onPlayFailed", "onPictureCaptured", "onDownloadProgress", "onDownloadSuccess", "onDownloadError", "onPlayerMessage", "onPlaybackProgress")

      // Imperative methods are now defined on the view, matching Android.
      AsyncFunction("capturePicture") { (view: ExpoEzvizView) in
        view.capturePicture()
      }

      AsyncFunction("startPlayback") { (view: ExpoEzvizView, recordFileDict: [String: Any]) -> Bool in
        return view.startPlayback(from: recordFileDict)
      }

      AsyncFunction("stopPlayback") { (view: ExpoEzvizView) -> Bool in
        return view.stopPlayback()
      }

      AsyncFunction("startLocalRecord") { (view: ExpoEzvizView, path: String) -> Bool in
        return view.startLocalRecord(with: path)
      }

//      AsyncFunction("stopLocalRecord") { (view: ExpoEzvizView) -> Void in
//        view.stopLocalRecord()
//      }

      AsyncFunction("downloadRecordFile") { (view: ExpoEzvizView, recordFileDict: [String: Any]) in
        view.downloadRecordFile(from: recordFileDict)
      }

      AsyncFunction("openSound") { (view: ExpoEzvizView) in
        view.openSound()
      }

      AsyncFunction("closeSound") { (view: ExpoEzvizView) in
        view.closeSound()
      }

      AsyncFunction("startRealPlay") { (view: ExpoEzvizView) in
        view.startRealPlay()
      }

      AsyncFunction("stopRealPlay") { (view: ExpoEzvizView) in
        view.stopRealPlay()
      }

      AsyncFunction("pausePlayback") { (view: ExpoEzvizView) -> Bool in
        return view.pausePlayback()
      }

      AsyncFunction("resumePlayback") { (view: ExpoEzvizView) -> Bool in
        return view.resumePlayback()
      }

      AsyncFunction("seekPlayback") { (view: ExpoEzvizView, offsetTimestamp: Double) in
        view.seekPlayback(to: offsetTimestamp)
      }
    }
  }

}

File: android/src/main/java/com/poseidon/ExpoEzvizModule.kt
package com.poseidon

import android.os.Build
import androidx.annotation.RequiresApi
import android.util.Log
import com.videogo.exception.BaseException
import com.videogo.openapi.EZConstants
import com.videogo.openapi.EZGlobalSDK
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.time.Duration
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.GregorianCalendar

class ExpoDeviceInfo : Record {
    @Field var cameraNum: Int = 0
    @Field var category: String? = null
    @Field var defence: Int = 0
    @Field var detectorNum: Int = 0
    @Field var deviceCover: String? = null
    @Field var deviceName: String? = null
    @Field var deviceSerial: String? = null
    @Field var deviceType: String? = null
    @Field var deviceVersion: String? = null
    @Field var isEncrypt: Boolean = false
    @Field var status: Int = 0
}

class ExpoDeviceRecordFile : Record {
    @Field var startTime: Long = 0
    @Field var stopTime: Long = 0
}

class ExpoEzvizModule : Module() {
  @RequiresApi(Build.VERSION_CODES.O)
  override fun definition() = ModuleDefinition {
    Name("ExpoEzviz")

    AsyncFunction("initSDK") { appKey: String, apiUrl: String? ->
      val application = appContext.activityProvider?.currentActivity?.application
        ?: throw IllegalStateException("Application context is not available.")

      EZGlobalSDK.showSDKLog(true)
      EZGlobalSDK.setDebugStreamEnable(true)
      EZGlobalSDK.enableP2P(false)
      EZGlobalSDK.initLib(application, appKey)

      if (apiUrl != null && apiUrl.isNotEmpty()) {
          EZGlobalSDK.getInstance().setServerUrl(apiUrl, "")
      }
    }

    AsyncFunction("setAccessToken") { accessToken: String ->
      print("EzvizModule: Setting Access Token")
      EZGlobalSDK.getInstance().setAccessToken(accessToken)
    }

    AsyncFunction("searchRecordFileFromDevice") { deviceSerial: String, cameraNo: Int, beginTimeString: String, endTimeString: String, promise: Promise ->
      val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss").withZone(ZoneOffset.UTC)

      val beginTime = try { ZonedDateTime.parse(beginTimeString, formatter) } catch (e: Exception) { null }
      val endTime = try { ZonedDateTime.parse(endTimeString, formatter) } catch (e: Exception) { null }

      if (beginTime == null || endTime == null) {
        promise.reject("date-format-error", "Could not parse date strings. Expected format: yyyy-MM-dd HH:mm:ss", null)
        return@AsyncFunction
      }

      val timeIntervalSeconds = Duration.between(beginTime, endTime).seconds
      if (timeIntervalSeconds > 24 * 60 * 60) {
        promise.reject("query-window-error", "The time window for querying records cannot exceed 24 hours.", null)
        return@AsyncFunction
      }

      val beginCalendar = GregorianCalendar.from(beginTime)
      val endCalendar = GregorianCalendar.from(endTime)

      Thread {
        try {
          val serializableRecords = EZGlobalSDK.getInstance()
            .searchRecordFileFromDevice(deviceSerial, cameraNo, beginCalendar, endCalendar)
            .map { record ->
                ExpoDeviceRecordFile().apply {
                    this.startTime = record.startTime?.timeInMillis ?: 0
                    this.stopTime = record.stopTime?.timeInMillis ?: 0
                }
            }
          promise.resolve(serializableRecords)
        } catch (e: BaseException) {
          promise.reject("search-error", e.localizedMessage, e)
        }
      }.start()
    }

    AsyncFunction("addDevice") { deviceSerial: String, verifyCode: String, promise: Promise ->
      Thread {
        try {
          EZGlobalSDK.getInstance().addDevice(deviceSerial, verifyCode)
          promise.resolve(null)
        } catch (e: BaseException) {
          promise.reject("add-device-error", e.localizedMessage, e)
        }
      }.start()
    }

    AsyncFunction("deleteDevice") { deviceSerial: String, promise: Promise ->
      Thread {
        try {
          EZGlobalSDK.getInstance().deleteDevice(deviceSerial)
          promise.resolve(null)
        } catch (e: BaseException) {
          promise.reject("delete-device-error", e.localizedMessage, e)
        }
      }.start()
    }

    AsyncFunction("setDeviceName") { deviceName: String, deviceSerial: String, promise: Promise ->
      Thread {
        try {
          EZGlobalSDK.getInstance().setDeviceName(deviceSerial, deviceName)
          promise.resolve(null)
        } catch (e: BaseException) {
          promise.reject("set-device-name-error", e.localizedMessage, e)
        }
      }.start()
    }

    AsyncFunction("setDeviceEncryptStatus") { deviceSerial: String, verifyCode: String, encryptStatus: Boolean, promise: Promise ->
      Thread {
        try {
          EZGlobalSDK.getInstance().setDeviceEncryptStatus(deviceSerial, verifyCode, encryptStatus)
          promise.resolve(null)
        } catch (e: BaseException) {
          promise.reject("set-encrypt-status-error", e.localizedMessage, e)
        }
      }.start()
    }

    AsyncFunction("setDefence") { deviceSerial: String, defenceType: Int, promise: Promise ->
      val defenceStatus = when (defenceType) {
        0 -> EZConstants.EZDefenceStatus.EZDefence_IPC_CLOSE
        1 -> EZConstants.EZDefenceStatus.EZDefence_IPC_OPEN
        8 -> EZConstants.EZDefenceStatus.EZDefence_ALARMHOST_ATHOME
        16 -> EZConstants.EZDefenceStatus.EZDefence_ALARMHOST_OUTER
        else -> null
      }

      if (defenceStatus == null) {
        promise.reject("invalid-defence-type", "Invalid defence type. Use 0 (Off/Sleep), 1 (On), 8 (At Home), or 16 (Outer).", null)
        return@AsyncFunction
      }

      Thread {
        try {
          EZGlobalSDK.getInstance().setDefence(deviceSerial, defenceStatus)
          promise.resolve(null)
        } catch (e: BaseException) {
          promise.reject("set-defence-error", e.localizedMessage, e)
        }
      }.start()
    }

    AsyncFunction("getDeviceInfo") { deviceSerial: String, promise: Promise ->
      Thread {
        try {
          val deviceInfo = EZGlobalSDK.getInstance().getDeviceInfo(deviceSerial)
          Log.d("ExpoEzvizModule", "deviceInfo: $deviceInfo")
          val expoDeviceInfo = ExpoDeviceInfo().apply {
            this.cameraNum = deviceInfo.cameraNum
            this.category = deviceInfo.category
            this.defence = deviceInfo.defence
            this.detectorNum = deviceInfo.detectorNum
            this.deviceCover = deviceInfo.deviceCover
            this.deviceName = deviceInfo.deviceName
            this.deviceSerial = deviceInfo.deviceSerial
            this.deviceType = deviceInfo.deviceType
            this.deviceVersion = deviceInfo.deviceVersion
            this.isEncrypt = deviceInfo.isEncrypt == 1
            this.status = deviceInfo.status
          }
          promise.resolve(expoDeviceInfo)
        } catch (e: BaseException) {
          promise.reject("get-device-info-error", e.localizedMessage, e)
        }
      }.start()
    }

    View(ExpoEzvizView::class) {
      Prop("deviceSerial") { view: ExpoEzvizView, deviceSerial: String? ->
        view.deviceSerial = deviceSerial
      }

      Prop("cameraNo") { view: ExpoEzvizView, cameraNo: Int ->
        view.cameraNo = cameraNo
      }

      Prop("verifyCode") { view: ExpoEzvizView, verifyCode: String? ->
        view.verifyCode = verifyCode
      }

      Prop("autoplay") { view: ExpoEzvizView, autoplay: Boolean? ->
        view.autoplay = autoplay ?: false
      }

      Prop("defaultSoundOn") { view: ExpoEzvizView, defaultSoundOn: Boolean? ->
        view.setDefaultSoundOn(defaultSoundOn)
      }

      Events("onLoad", "onPlayFailed", "onPictureCaptured", "onDownloadProgress", "onDownloadSuccess", "onDownloadError", "onPlayerMessage", "onPlaybackProgress")

      AsyncFunction("capturePicture") { view: ExpoEzvizView ->
        view.capturePicture()
      }

      AsyncFunction("startPlayback") { view: ExpoEzvizView, recordFileDict: Map<String, Any> ->
        return@AsyncFunction view.startPlayback(recordFileDict)
      }

      AsyncFunction("stopPlayback") { view: ExpoEzvizView ->
        return@AsyncFunction view.stopPlayback()
      }

      AsyncFunction("stopLocalRecord") { view: ExpoEzvizView ->
        view.stopLocalRecord()
      }

      AsyncFunction("downloadRecordFile") { view: ExpoEzvizView, recordFileDict: Map<String, Any> ->
        view.downloadRecordFile(recordFileDict)
      }

      AsyncFunction("openSound") { view: ExpoEzvizView ->
        view.openSound()
      }

      AsyncFunction("closeSound") { view: ExpoEzvizView ->
        view.closeSound()
      }

      AsyncFunction("startRealPlay") { view: ExpoEzvizView ->
        view.startRealPlay()
      }

      AsyncFunction("stopRealPlay") { view: ExpoEzvizView ->
        view.stopRealPlay()
      }

      AsyncFunction("pausePlayback") { view: ExpoEzvizView ->
        return@AsyncFunction view.pausePlayback()
      }

      AsyncFunction("resumePlayback") { view: ExpoEzvizView ->
        return@AsyncFunction view.resumePlayback()
      }

      AsyncFunction("seekPlayback") { view: ExpoEzvizView, offsetTimestamp: Double ->
        return@AsyncFunction view.seekPlayback(offsetTimestamp)
      }

      // Renamed to match iOS for a unified API
      AsyncFunction("startLocalRecord") { view: ExpoEzvizView, path: String ->
        return@AsyncFunction view.startLocalRecordWithFile(path)
      }
    }
  }
}
